Improve password update functionality and add tests

avifenesh opened the pull request:
Signed-off-by: avifenesh <aviarchi1994@gmail.com>

<!--
Thanks for contributing to Valkey GLIDE!

Please make sure you are aware of our contributing guidelines [available
here](https://github.com/valkey-io/valkey-glide/blob/main/CONTRIBUTING.md)

-->

Enhance the `update_connection_password` method to support re-authentication while disconnected and add tests to verify the improvements.

### Issue link

This Pull Request is linked to issue https://github.com/valkey-io/valkey-glide/issues/2360

### Checklist

Before submitting the PR make sure the following are checked:

-   [x] This Pull Request is related to one issue.
-   [x] Commit message has a detailed description of what changed and why.
-   [x] Tests are added or updated.
-   [x] CHANGELOG.md and documentation files are updated.
-   [x] Destination branch is correct - main or release
-   [x] Commits will be squashed upon merging.

Enhance the `update_connection_password` method to support re-authentication and add tests to verify the improvements.

Yury-Fridlyand commented on glide-core/src/client/mod.rs:
```diff
@@ -485,24 +485,24 @@ impl Client {
+        }
         if re_auth {
             let routing = RoutingInfo::MultiNode((
                 MultipleNodeRoutingInfo::AllNodes,
```
Good to add to docs that client re-auth on all nodes

avifenesh replied:
I'll take care of it. Do you think it makes sense that it would do different? Seems clear to me that this is the impl.

Yury-Fridlyand replied:
That's obvious, but good to have

avifenesh replied:
Done.

Yury-Fridlyand commented on node/src/BaseClient.ts:
```diff
@@ -7691,7 +7691,7 @@ export class BaseClient {
      * @param password - The new password to update the current password, or `null` to remove the current password.
      * @param reAuth - If `true`, the client will re-authenticate immediately with the new password. If `false`, the new password will be used for the next connection attempt.
-     * @returns Always `"OK"`.
+     * @returns `"Ok"`, if `reAauth: true` returns `"OK"` if the reauthenticate succeed.
```
What if reauth = false?

avifenesh replied:
@Yury-Fridlyand note: I removed completely the “will be used in the next” as I understand it creates confusion. It is a binary flag.

Yury-Fridlyand reviewed (commented):
GLIDE core still sends HELLO without AUTH

Yury-Fridlyand commented on python/python/glide/async_commands/core.py:
```diff
@@ -415,11 +415,10 @@ async def update_connection_password(
             re_auth (bool):
-                - `True`: The client will re-authenticate immediately with the new password.
-                - `False`: The new password will be used for the next connection attempt.
+                - `True`: The client will re-authenticate immediately with the new password against all connections.
```
Same - we still need to desctibe what `False` is doing

Yury-Fridlyand commented on package.json:
```diff
@@ -3,7 +3,7 @@
         "@types/eslint__js": "^8.42.3",
         "@types/eslint-config-prettier": "^6.11.3",
-        "eslint": "^9.10.0",
+        "eslint": "9.14.0",
```
This should probably go in another PR ahead of all others.

avifenesh replied:
If all ready to merge it, lets merge, otherwise ill open seperatly. but i actually hope you didn't waited for me to fix the CI with one char removal.

avifenesh commented on node/tests/SharedTests.ts:
```diff
@@ -12287,6 +12287,20 @@ export function runBaseTests(config: {
                         const value2 = await client.get("test_key2");
                         expect(value2).toEqual("test_value2");
+                        // Testing that after kill update password works with `reAuth: true` works
+                        await client.customCommand([
```
I move it, as in python to seperate test suite. The AUTH is a whole module of different functionalities and im creating the pltform for it with the feature, making it ready for next tasks.
As part of another suite it will close the the cluster after finishing.
You should do the same in your PR.

avifenesh commented on node/src/BaseClient.ts:
```diff
@@ -7690,8 +7690,7 @@ export class BaseClient {
      * @param password - The new password to update the current password, or `null` to remove the current password.
-     * @param reAuth - If `true`, the client will re-authenticate immediately with the new password. If `false`, the new password will be used for the next connection attempt.
-     * @returns Always `"OK"`.
+     * @param reAuth - If `true`, the client will re-authenticate immediately with the new password against all connections.
```
Same response as in Python.

avifenesh replied:
Im resolving comments that i unswer on for my arrangment. If you see it as need reopen and keep discuess feel free to do so.
If disussion get into very small details, please lets merge so we good to release and keep discuess enhancemnt on a seperate PR.

avifenesh commented:
@Yury-Fridlyand fixed misbehavior, added proper error message on timeout, fixed and added tests and proper teardown. Changed doc to be more clear. Please rebase and check your tests, approve if all pass.
Align code to python and node logic.

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                        }
+
+                        // Update password without re-authentication
+                        const result = await client.updateConnectionPassword(
```
i.e. 
```
                        await client.configSet({ requirepass: "" });
```

acarbonetto commented on node/src/BaseClient.ts:
```diff
@@ -7689,20 +7689,24 @@ export class BaseClient {
-     * @returns Always `"OK"`.
+     * @param password - `String | null`. The new password to update the current password, or `null` to remove the current password.
+     * @param immidiateAuth - A `boolean` flag. If `true`, the client will authenticate immediately with the new password against all connections, Using `AUTH` command.
+     *                 If password suplied is an empty string, auth will not be perform and warning will be rturned.
```
```suggestion
     *                 If password supplied is an empty string, auth will not be perform and warning will be returned.
```

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+             */
+            it("test_update_connection_password_no_server_auth", async () => {
+                await runTest(async (client: BaseClient) => {
+                    try {
```
its not clear whether this test is testing: 
1. if the client does not require password, OR 
2. if the client's password is not NEW_PASSWORD (ie wrong password auth failure)

acarbonetto replied:
according to your test documentation, you want it to test the first case, and we should make sure that's the case with the configset call above.

avifenesh replied:
Setup been answered above. Assume all tests starts with no auth require.

I'm testing what will happen if you take a non-existent password and ask the client to perform re-auth with it.
I expect it to error.

For the below.
The client is fine, he can update again.
If it was missed, the update feature, update config, not the server code. Auth from config is needed just on connection. The failure above doesn't disconnect, it is just the command that failed.

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                                false,
+                            ),
+                        ).resolves.toBe("OK");
+                        await client.configSet({
```
what is this doing?

acarbonetto replied:
doesn't this call fail because the password isn't correct?

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                        await client.updateConnectionPassword(
+                            NEW_PASSWORD,
+                            true,
+                        );
```
expect "OK" here

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                        );
+                        await client.configSet({
+                            requirepass: "",
+                        });
```
expect "OK" here

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+            /**
+             * Test changing server password when connection is lost before password update.
+             * Verifies that the client will not be able to reach the inner core and return an error.
+             */
```
Maybe describe this case a little better and point out why this only happens with the standalone client?  Seems like for a standalone client, the 'inner core' makes a request to redis (which fails) and thus the updateConnectionPassword throws a RequestError. 
... but I'm just guessing and it would be nice to have this explained somewhere.

avifenesh replied:
Ill describe more.

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                    ]);
+                    // Try updating client password without immediate re-auth
+                    await expect(
+                        client.updateConnectionPassword(NEW_PASSWORD, false),
```
if this passes `true` does this command work?  
should we test the `true` case too?

avifenesh replied:
I'll add also. But it will fail. There is no one listening on the other side to send the command.

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+             * Test replacing connection password with immediate re-authentication using a non-valid password.
+             * Verifies that immediate re-authentication fails when the password is not valid.
+             */
+            it("test_update_connection_password_auth_non_valid_pass", async () => {
```
can you put this test above?  
It would be nice to have all the 'simple' and 'sunny-day' (success) test cases at the top of the file

avifenesh replied:
haha this is the sunny-day? both fails. but will do.

acarbonetto commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+    describe.each([ProtocolVersion.RESP2, ProtocolVersion.RESP3])(
+        "update_connection_password_%p",
+        (protocol) => {
+            const NEW_PASSWORD = "new_password";
```
I would recommend using a random password for each test case - to make sure that tests don't overlap/interfere

avifenesh replied:
I'm afraid I'm using the same one for the same reason, but opposite, you want me not to.
I want to be able, even if a test fails, to recover the server for the whole other tests.
Otherwise, if my tests are failing, it will fail all tests until recreation of the server.

avifenesh commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+
+    afterAll(async () => {
+        if (managementClient) {
+            managementClient.close();
```
Absolutely right.

avifenesh replied:
Well, no, didn't noticed, its another client which stay alive through the run.

avifenesh commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                        const value2 = await client.get("test_key2");
+                        expect(value2).toEqual("test_value2");
+
+                        // Clean up - revert server password
```
Im confuse by my own code now

avifenesh replied:
IDK its not needed, i added after each, ill remove all of those.

avifenesh commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                await runTest(async (client: BaseClient) => {
+                    try {
+                        await expect(
+                            client.updateConnectionPassword(NEW_PASSWORD, true),
```
answered above

avifenesh commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,339 @@
+                  getServerVersion,
+              )
+            : await ValkeyCluster.createCluster(true, 3, 1, getServerVersion);
+        managementClient = standaloneAddresses
```
We actually need both, its not an if. if you run your own specific cluster than the condition is relevant.

acarbonetto commented on node/src/BaseClient.ts:
```diff
@@ -7693,20 +7693,24 @@ export class BaseClient {
-     * @returns Always `"OK"`.
+     * @param password - `String | null`. The new password to update the current password, or `null` to remove the current password.
+     * @param immidiateAuth - A `boolean` flag. If `true`, the client will authenticate immediately with the new password against all connections, Using `AUTH` command.
+     *                 If password supplied is an empty string, auth will not be perform and warning will be returned.
```
```suggestion
     *                 If password supplied is an empty string, the client will not perform auth and instead a warning will be returned.
```
```suggestion
     *                 If password supplied is an empty string, the client will not perform auth and a warning will be returned.
```

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+    });
+
+    afterAll(async () => {
+        if (cmdCluster) {
+            await cmdCluster.close();
+        }
+
+        if (cmeCluster) {
+            await cmeCluster.close();
+        }
+
+        if (managementClient) {
+            managementClient.close();
+        }
```
```suggestion
        await cmdCluster?.close();
        await cmeCluster?.close();
        managementClient?.close();
```
Do we need `await` for the last `close`?

avifenesh replied:
no need

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+        protocol: ProtocolVersion,
+        configOverrides?: Partial<BaseClientConfiguration>,
+    ) => {
+        const isStandaloneMode = configOverrides?.addresses?.length === 1;
```
We can run cluster and create config with only 1 address

Yury-Fridlyand replied:
I think this happens when you use `--cluster-endpoints`

avifenesh replied:
Yea, its dealt in setup of all tests then above

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+                        await client.configSet({ requirepass: NEW_PASSWORD });
+
+                        // Update client password with re-auth
+                        await expect(
+                            client.updateConnectionPassword(NEW_PASSWORD, true),
+                        ).resolves.toBe("OK");
```
```suggestion
                        expect(await
                            client.updateConnectionPassword(NEW_PASSWORD, true),
                        ).toEqual("OK");
```

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+                        "TYPE",
+                        "normal",
+                    ]);
+                    // Try updating client password without immediate re-auth and with, both should fail
```
Is that the same I reported on GH? If yes, post here links, so we can easily update the tests once fixed.

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+             */
+            it("test_update_connection_password_connection_lost_before_password_update", async () => {
+                await runTest(async (client: BaseClient) => {
+                    if (client instanceof GlideClusterClient) {
```
Update description or add a comment why you're testing on stadalone only

avifenesh replied:
The explanation is pretty long, all those details are not enough?

avifenesh replied:
added another line

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+                "test_update_connection_password",
+                async () => {
+                    await runTest(async (client: BaseClient) => {
+                        if (client instanceof GlideClient) {
```
Why skipping stadalone?

avifenesh replied:
```
             * Currently, this test is only supported for cluster mode,
             * since standalone mode dont have multiple connections to manage,
             * and the client will try to reconnect and will not listen to new tasks.
    ```

Yury-Fridlyand replied:
Oh shame on me

Yury-Fridlyand commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+            it("test_update_connection_password_long", async () => {
+                await runTest(async (client: BaseClient) => {
+                    const longPassword = "p".repeat(1000);
+                    await expect(
+                        client.updateConnectionPassword(longPassword, false),
+                    ).resolves.toBe("OK");
```
```suggestion
                    expect(
                        await client.updateConnectionPassword(longPassword, false),
                    ).toEqual("OK");
```

avifenesh commented on glide-core/redis-rs/afl/parser/Cargo.toml:
for whom use ARM machine

avifenesh commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+    afterEach(async () => {
+        if (managementClient) {
+            try {
+                await managementClient.customCommand(["AUTH", "new_password"]);
```
Will move the flushall down.

avifenesh commented on node/tests/AuthTest.test.ts:
```diff
@@ -0,0 +1,329 @@
+            }
+
+            try {
+                await client.updateConnectionPassword("");
```
Afterall, this each

avifenesh commented on python/python/tests/conftest.py:
```diff
@@ -268,17 +285,81 @@ async def create_client(
         return await GlideClient.create(config)
 
 
+NEW_PASSWORD = "new_secure_password"
+WRONG_PASSWORD = "wrong_password"
+
+
+async def auth_client(client: TGlideClient, password):
+    """
+    Authenticates the given TGlideClient server connected.
+    """
+    if isinstance(client, GlideClient):
+        await client.custom_command(["AUTH", password])
+    elif isinstance(client, GlideClusterClient):
+        await client.custom_command(["AUTH", password], route=AllNodes())
+
+
+async def config_set_new_password(client: TGlideClient, password):
+    """
+    Sets a new password for the given TGlideClient server connected.
+    This function updates the server to require a new password.
+    """
+    if isinstance(client, GlideClient):
+        await client.config_set({"requirepass": password})
+    elif isinstance(client, GlideClusterClient):
+        await client.config_set({"requirepass": password}, route=AllNodes())
+
+
+async def kill_connections(client: TGlideClient):
+    """
+    Kills all connections to the given TGlideClient server connected.
+    """
+    if isinstance(client, GlideClient):
+        await client.custom_command(["CLIENT", "KILL", "TYPE", "normal"])
+    elif isinstance(client, GlideClusterClient):
+        await client.custom_command(
+            ["CLIENT", "KILL", "TYPE", "normal"], route=AllNodes()
+        )
```
Im doing broader work on the python tests platform, moving a lot of stuff around, after release.

avifenesh commented on python/python/tests/conftest.py:
```diff
@@ -268,17 +285,81 @@ async def create_client(
+        )
+        await client.custom_command(["FLUSHALL"])
+        await client.close()
+    except ClosingError as e:
+        # Check if the error is due to authentication
+        if "NOAUTH" in str(e):
```
Thats for general error handling of optional problems that can happens if something got broken and i make all the suites fails because of this suite.

avifenesh commented on python/python/tests/test_auth.py:
```diff
@@ -1,126 +1,129 @@
+import asyncio
```
Always forget

avifenesh commented:
> Add this PR to change log please (you can update an existing line)

@Yury-Fridlyand done.
I'm answering and resolving, if something is not a sufficient answer or there's still disagreement, unresolve.
